service/history/execution/context_test.go

```diff
@@ -0,0 +1,939 @@
+					RangeID: 100,
+				}).Return(&persistence.CreateWorkflowExecutionResponse{
+					MutableStateUpdateSessionStats: &persistence.MutableStateUpdateSessionStats{
+						MutableStateSize: 123,
+					},
+				}, nil)
+			},
+			want: &persistence.CreateWorkflowExecutionResponse{
+				MutableStateUpdateSessionStats: &persistence.MutableStateUpdateSessionStats{
+					MutableStateSize: 123,
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Success case with retry",
+			request: &persistence.CreateWorkflowExecutionRequest{
+				RangeID: 100,
+			},
+			mockSetup: func(mockShard *shard.MockContext) {
+				mockShard.EXPECT().CreateWorkflowExecution(gomock.Any(), &persistence.CreateWorkflowExecutionRequest{
+					RangeID: 100,
+				}).Return(nil, &types.ServiceBusyError{})
+				mockShard.EXPECT().CreateWorkflowExecution(gomock.Any(), &persistence.CreateWorkflowExecutionRequest{
+					RangeID: 100,
+				}).Return(&persistence.CreateWorkflowExecutionResponse{
+					MutableStateUpdateSessionStats: &persistence.MutableStateUpdateSessionStats{
+						MutableStateSize: 123,
+					},
+				}, nil)
+			},
+			want: &persistence.CreateWorkflowExecutionResponse{
+				MutableStateUpdateSessionStats: &persistence.MutableStateUpdateSessionStats{
+					MutableStateSize: 123,
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "workflow already started",
+			request: &persistence.CreateWorkflowExecutionRequest{
+				RangeID: 100,
+			},
+			mockSetup: func(mockShard *shard.MockContext) {
+				mockShard.EXPECT().CreateWorkflowExecution(gomock.Any(), gomock.Any()).Return(nil, &persistence.WorkflowExecutionAlreadyStartedError{})
+			},
+			wantErr: true,
+			assertErr: func(t *testing.T, err error) {
+				assert.IsType(t, err, &persistence.WorkflowExecutionAlreadyStartedError{})
+			},
+		},
+		{
+			name: "timeout",
+			request: &persistence.CreateWorkflowExecutionRequest{
+				RangeID: 100,
+			},
+			mockSetup: func(mockShard *shard.MockContext) {
+				mockShard.EXPECT().CreateWorkflowExecution(gomock.Any(), gomock.Any()).Return(nil, &persistence.TimeoutError{})
+			},
+			wantErr: true,
+			assertErr: func(t *testing.T, err error) {
+				assert.IsType(t, err, &persistence.TimeoutError{})
+			},
+		},
+		{
+			name: "other non retryable error",
+			request: &persistence.CreateWorkflowExecutionRequest{
+				RangeID: 100,
+			},
+			mockSetup: func(mockShard *shard.MockContext) {
+				mockShard.EXPECT().CreateWorkflowExecution(gomock.Any(), gomock.Any()).Return(nil, errors.New("some error"))
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockShard := shard.NewMockContext(mockCtrl)
+			if tc.mockSetup != nil {
+				tc.mockSetup(mockShard)
+			}
+			resp, err := createWorkflowExecutionWithRetry(context.Background(), mockShard, testlogger.New(t), tc.request)
```

Comment:
I'd make the retry policy overridable so test can pass low value for max retry attempt/duration. 
You can then add a case which keeps failing even after all attempts